qiskit/extensions/quantum_initializer/initializer.py

```diff
@@ -13,326 +13,75 @@
-            )
-
-        for qubit, bit in enumerate(intstr):
-            initialize_circuit.append(Reset(), [q[qubit]])
-            if bit == "1":
-                initialize_circuit.append(XGate(), [q[qubit]])
-
-        return initialize_circuit
-
-    def _define_synthesis(self):
-        """Calculate a subcircuit that implements this initialization
-
-        Implements a recursive initialization algorithm, including optimizations,
-        from "Synthesis of Quantum Logic Circuits" Shende, Bullock, Markov
-        https://arxiv.org/abs/quant-ph/0406176v5
-
-        Additionally implements some extra optimizations: remove zero rotations and
-        double cnots.
+        Args:
+            params (str, list, int or Statevector):
+                * Statevector: Statevector to initialize to.
+                * list: vector of complex amplitudes to initialize to.
+                * string: labels of basis states of the Pauli eigenstates Z, X, Y. See
+                  :meth:`.Statevector.from_label`.
+                  Notice the order of the labels is reversed with respect to the qubit index to
+                  be applied to. Example label '01' initializes the qubit zero to :math:`|1\rangle`
+                  and the qubit one to :math:`|0\rangle`.
+                * int: an integer that is used as a bitmap indicating which qubits to initialize
+                  to :math:`|1\rangle`. Example: setting params to 5 would initialize qubit 0 and qubit 2
+                  to :math:`|1\rangle` and qubit 1 to :math:`|0\rangle`.
+
+            num_qubits (int): This parameter is only used if params is an int. Indicates the total
+                number of qubits in the `initialize` call. Example: `initialize` covers 5 qubits
+                and params is 3. This allows qubits 0 and 1 to be initialized to :math:`|1\rangle`
+                and the remaining 3 qubits to be initialized to :math:`|0\rangle`.
         """
-        # call to generate the circuit that takes the desired vector to zero
-        disentangling_circuit = self.gates_to_uncompute()
+        self._stateprep = StatePreparation(params, num_qubits)
 
-        # invert the circuit to create the desired vector from zero (assuming
-        # the qubits are in the zero state)
-        initialize_instr = disentangling_circuit.to_instruction().inverse()
+        super().__init__("initialize", self._stateprep.num_qubits, 0, self._stateprep.params)
 
+    def _define(self):
         q = QuantumRegister(self.num_qubits, "q")
         initialize_circuit = QuantumCircuit(q, name="init_def")
-        for qubit in q:
-            initialize_circuit.append(Reset(), [qubit])
-        initialize_circuit.append(initialize_instr, q[:])
-
-        return initialize_circuit
+        initialize_circuit.reset(q)
+        initialize_circuit.append(self._stateprep, q)
+        self.definition = initialize_circuit
 
     def gates_to_uncompute(self):
```

Comment:
Great catch! It uses it since it as a workaround to get the initialization without the resets -- which is exactly the state preparation we're adding. So yeah, let's not deprecate it yet and as soon as finance uses the state preparation we can deprecate.